---
title: "Class 15: Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

 ##Import countData and colData
 (make factors false (true is default) because dont want that as headings)
 
```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```
## dim is to find out no. of rows and columns and nrow and rcol is to separately count
```{r}
dim(counts)
head(counts)
```



```{r}
head(metadata)
```


#  have to check if ^ and bottom table columns (first?) and rows of bottom table matches 
```{r}
colnames(counts)
```

```{r}
metadata$id
```

# Check for correspondence of metadata and countdata (check for *all*)

```{r}
all(colnames(counts)[-1] == metadata$id)
```

## First task: find which columns in **counts** correspond to the control conditions (i.e. no drug!). Then we can calculate the mean count value for each gene.

```{r}
control.inds <- metadata$dex == "control"
control.ids <- metadata$id[control.inds]

head(counts[,control.ids])
```
 ## Calc. the mean counts value for all genes
 
```{r}
control.mean <- rowSums(counts[,control.ids]) / ncol(counts[,control.ids])
names(control.mean) <- counts$ensgene
head(control.mean)
```

## Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.inds <- metadata$dex == "treated"
treated.ids <- metadata$id[treated.inds]

head(counts[,treated.ids])
```

```{r}
treated.mean <- rowSums(counts[,treated.ids]) / ncol(counts[,treated.ids])
treated.mean
```

```{r}
names(treated.mean) <- counts$ensgene
head(treated.mean)
```

## combine our meancount data into a new data frame called meancounts!

```{r}
meancounts <- data.frame(control.mean, treated.mean)
meancounts
```

```{r}
colSums(meancounts)
```

```{r}
plot(meancounts$treated.mean, meancounts$control.mean)
```

## ^ plot shows effect, but all clustered in the beginning so better to use a log log plot:

```{r}
plot(meancounts$treated.mean, meancounts$control.mean, log = "xy")
```

We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the head() or the View() function for example.

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

## Filter out these cases with 0 as the result so you don't get weird results like Inf and Nan

```{r}
toy <- c(1,2,0,4,0)
toy == 0
which(toy == 0)
```

```{r}
toy <- meancounts[1:6,]
toy
toy[,1:2]
toy[,1:2] == 0
which(toy[,1:2] == 0, arr.ind=TRUE)
unique(which(toy[,1:2] == 0, arr.ind=TRUE)[,1])
```

# (^ always check back w/ above data to see if results match!)


Remove the Zero entry genes from meancounts to produce **my counts**

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
sum(up.ind)
```

```{r}
sum(down.ind)
```


There are `sum(up.ind)` upregulated genes and `sum(down.ind)` down regulated genes. 
##(back ticks ` run the function in R code!)

```{r}
head(mycounts[up.ind,])
```

## Adding annotation data

Use the **merge ()** function to add annotation data from ensemble


```{r}
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```

```{r}
mycounts.annotated <- merge(x=mycounts, y=anno, by.x="row.names", by.y = "ensgene")
head (mycounts.annotated)
```

## DeSeq2 analysis

Now we do this analyses properly with p-values!

```{r}
library(DESeq2)
citation("DESeq2")
```

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```


```{r}
dds <- DESeq(dds)
```
Get the results:
```{r}
res <- results(dds)
head(res)
```

```{r}
summary(res)
```

```{r}
res01 <- results(dds, alpha=0.01)
summary(res01)
```

# Data Visualization:

```{r}
plot(res01$log2FoldChange, -log(res01$padj))
```

```{r}
mycols <- rep("gray", nrow(res01))
mycols[abs(res01$log2FoldChange) > 2] <- "red"

inds <- (res01$padj < 0.01) & (abs(res01$log2FoldChange)>2)
mycols[inds] <- "blue"

##Volcano plot with custom colors:
plot(res01$log2FoldChange, -log(res01$padj), col=mycols, ylab = "-Log(P-value)", xlab = "Log2(FoldChange)")

#optional?
abline(v=c(-2,2), col = "gray", lty = 2)
abline(h=-log(0,1), col="gray", lty = 2)
```














