---
title: "Class 18: Cancer Genomics"
output: github_document
---

## NCI Genomic Data Commons

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

```{r}
GenomicDataCommons::status()
```

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

Q9. Write the R code to make a barplot of the cases per project. Lets plot this data with a log scale for the y axis (log="y"), rotated axis labels (las=2) and color the bar coresponding to the TCGA-PAAD project.

```{r}
x <- cases_by_project$project.project_id
head(x)
barplot(x$doc_count, names.arg = x$key, las = 2, log = "y")
```


```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
```

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```

```{r}
plotmafSummary(vars)
```

```{r}
lollipopPlot(vars, gene='TP53')
```


## Hands on Part 2: Designing a personalised cancer vaccine

Q1: Identify sequence regions that contain all 9-mer peptides that are only found in the tumor

```{r}
library(bio3d)
seqs <- read.fasta("~/Downloads/lecture18_sequences.fa")
seqs
```

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 


## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```


We can use these indices in `mutant.sites` to extract subsequences as required for the hands-on session. First however we come up with suitable names for these subsequences based on the mutation. This will help us later to make sense and keep track of our results.

```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

Now lets extract all 9-mer mutant encompassing sequences for each mutant site. This is equivalent to finding the sequence region eight residues before and eight residues after our mutation sites and outputting this subsequence to a new FASTA file.

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```

SO SUBSEQUENCES ANALYSIS:

best binding affinity is YLDDRNTFV peptide (HLA-A*02:01)

## Q3 - find something identical to this peptide?

?? why but answer is all are good to use for vaccine

 FVHSVVVPY possibly good because repeat so two chances w/ that

	SPLPSQAML best??








